besmarts-core/python/examples/07_besmarts_fit.py

```diff
@@ -100,54 +127,70 @@ def make():
-                gdt.values.extend(gx)
-            tid = assignments.ENERGY
-            if tid in rdata:
-                gdt = assignments.graph_db_table(topology.null)
-                fn = rdata[tid]
-                ene = [*map(float,
-                    [x for x in open(fn).read().split('\n') if x]
-                )]
-                gdt.values.extend(ene)
-                gde.tables[tid] = gdt
+        for rdata_dict in fn_list_of_dicts:
+            # === first add positions ===
+            positions_index: int = assignments.POSITIONS  # the index of the POSITIONS property in a graph_db_entry
+            # create new empty data table
+            positions_gdt = assignments.graph_db_table(topology.atom)
+            # create empty graph_db_graph. This holds data associated with a set of graphs
+            positions_gdg = assignments.graph_db_graph()
+            # add the graph_db_graph to the table of positions data
+            positions_gdt.graphs[molecule_graph_id] = positions_gdg
+            # read the positions xyz data from the dictionary
+            positions_gdr: graphs.graph_db_row = load_xyz([rdata_dict[positions_index]])
+            # add the positions data to the graph_db_graph
+            positions_gdg.rows[0] = positions_gdr
+            # add the table to the graph_db_entry
+            gde.tables[positions_index] = positions_gdt
+            
+            # === check if gradient data is in the dictionary ===
+            gradient_index: int = assignments.GRADIENTS
+            if gradient_index in rdata_dict:
+                gradient_gdt = assignments.graph_db_table(topology.atom)
+                gradient_gdg = assignments.graph_db_graph()
+                gradient_gdt.graphs[molecule_graph_id] = gradient_gdg
+                gradient_gdr = load_xyz([rdata_dict[gradient_index]])
+                gradient_gdg.rows[0] = gradient_gdr
+                gde.tables[gradient_index] = gradient_gdt
+                gradient_graph_db_column = gradient_gdr[0]
+                flat_values: list[float] = [
+                    gradient
+                    for gradient_list in gradient_graph_db_column.selections.values()
+                    for gradient in gradient_list
+                ]
+                gradient_gdt.values.extend(flat_values)
+
+            # === check if energy data is in the dictionary ===
+            energy_index: int = assignments.ENERGY
+            if energy_index in rdata_dict:
+                energy_gdt = assignments.graph_db_table(topology.null)
+                energy_data_file = rdata_dict[energy_index]
+                # assume energy data is a file path
+                with open(energy_data_file, 'r') as f:
+                    energy_data = f.read().split("\n")
+                # assume energy data is a list of floats with one energy per line
+                energy_values = [*map(float, [x for x in energy_data if x])]
+                energy_gdt.values.extend(energy_values)
```

Comment:
If this example contained multiple observations to fit to (so multiple gids I assume), how would besmarts map each energy value to the corresponding coordinates? I could be missing something here but I don't see any modification of `energy_gdt.graphs` (although I guess energies aren't actually part of this example!)

Reply:
So I never really intended to use energies in this example, and this snippet is more destructive than constructive. The value member field was just a lazy way to append data without having to deal with the data structure as I worked on various objective functions (hessian freqs, DLC gradients, etc), so I consider the values member to be "user defined" :). To answer your question generally, the (proper/designed) way to do this would be to set up a graph_db table (with a null topology since it is per graph/molecule/system) and then you would assign energies as you would like. I note that a weakness here is that I don't have a great way to set a total energy anywhere, it would have to be a sum over all graph_db_graph objectives for a particular row_id. This would give you a bunch of graph_db_columns, and for null topologies I reserve the selection (0,) to refer to the entire molecule, remembering that I use 1-index for selections for atoms so 0 is never used anywhere else. (I haven't actually done this anywhere, but this is how I will do it once I start fitting energies). This would be mostly for QM energies, but it should be possible to do IC energies instead for MM. One could set up a custom addressing to store the totals in the values member; this greatly simplifies computing the objective at the end, which is essentially why I added it in the first place.